src/extension/kernel.ts

```diff
@@ -310,47 +307,83 @@ export class Kernel implements Disposable {
     } else if (message.type === ClientMessages.githubWorkflowDispatch) {
       await handleGitHubMessage({ messaging: this.messaging, message })
     } else if (message.type === ClientMessages.displayPrompt) {
-      const answer = await window.showInputBox({
-        password: message.output.isSecret,
-        placeHolder: message.output.placeholder,
-        title: message.output.title,
-      })
       const cell = await getCellByUuId({ editor, uuid: message.output.uuid })
       if (!cell || message.output.uuid !== cell.metadata?.['runme.dev/uuid']) {
         return
       }
-      postClientMessage(this.messaging, ClientMessages.onPrompt, {
-        answer,
-        uuid: message.output.uuid,
-      })
+      const categories = await getNotebookCategories(this.context, cell.document.uri)
+
+      const input = window.createQuickPick<CategoryQuickPickItem>()
+      input.title = message.output.title
+      input.placeholder = message.output.placeholder
+      input.canSelectMany = true
+
+      const origCategories = categories.map((val) => new CategoryQuickPickItem(val))
+      input.items = origCategories
+      input.show()
+
+      this.#disposables.push(
+        input.onDidChangeValue((value) => {
+          const isNewItem = categories.filter((c) => c.includes(value)).length === 0
+          if (!isNewItem) {
+            input.items = origCategories
+            return
+          }
+
+          input.items = [
+            ...input.items.filter((i) => !i.isNew()),
+            new CategoryQuickPickItem(value, true),
+          ]
+
+          /**
+           * auto select new category if label is valid
+           */
+          if (CategoryQuickPickItem.isValid(value)) {
+            input.selectedItems = input.items.slice(-1)
+          }
+        }),
+        input.onDidChangeSelection((items) => {
+          if (items.length === 1 && !items[0].isValid()) {
+            input.selectedItems = []
+          }
+        }),
+        input.onDidAccept(() => {
+          /**
+           * validate new category label
+           */
+          if (input.selectedItems.length === 1 && !input.selectedItems[0].isValid()) {
+            return
+          }
+
+          input.dispose()
+          postClientMessage(this.messaging, ClientMessages.onPrompt, {
+            answer: input.selectedItems.map((qp) => qp.label).join(CATEGORY_SEPARATOR),
+            uuid: message.output.uuid,
+          })
+        }),
+      )
```

Comment:
I would like to start moving out logic from the Kernel, and have their own separate file for messages logic, take a look at messages folder, this could be inside `messages/categorySelector` or similar, WDYT?